### PR TITLE
Error at compile-time when a non-subclassable class is being subclassed

### LINKED ISSUE #4451: Detect non-subclassable class being subclassed at compile-time
The following compiles, but crashes at runtime:
```rust
#[pyclass] // No `(subclass)`!
pub struct BaseClass {}

#[pyclass(extends=BaseClass)]
pub struct SubClass {}
```
It'd be nice if we could detect this at compile-time instead. This should be simple: make a bound that isn't true for non-subclassable classes, possibly even with a nice `#[diagnostics::on_unimplemented]`.

I can work on this later.

### PR DESCRIPTION
**Pull request by ChayimFriedman2**
Previously this crashed at runtime.

We even get to remove few tests. Nothing more fun than removing a no-longer-needed test!

Fixes #4451.

### CODE REVIEW

**Review by davidhewitt — CHANGES_REQUESTED** [review]
Thanks very much for improving this; it's great to improve UX!

I think we might be able to keep the codebase a bit simpler by improving the existing `PyClassBaseType` trait instead of adding a new trait & generated code. At the moment all `#[pyclass]` types unconditionally implement it via a blanket impl, and most native types do via the `pyobject_native_type_sized!` macro. I guess we could add new bounds on the blanket, or even remove it and just emit implementations in the `#[pyclass]` macro if that improves diagnostics.

**Comment by ChayimFriedman2** [comment]
@davidhewitt Done.

**Comment by davidhewitt** [comment]
Overall this looks good to me, however there seem to be some feature combinations which aren't satisfied. I'd suggest running `nox -s check-feature-powerset` to help identify the broken cases and add `#[cfg]`s on them.

(Note that in the long run we should be able to remove a lot of the cfgs if we solved #1344. Which I think is very doable but just hasn't managed to top my priorities...

**Comment by ChayimFriedman2** [comment]
@davidhewitt Fixed that. Those cfgs are a mess!

File: tests/test_compile_error.rs [hunk 1]
```diff
@@ -62,4 +62,5 @@ fn test_compile_errors() {
     #[cfg(all(Py_LIMITED_API, not(Py_3_9)))]
     t.compile_fail("tests/ui/abi3_dict.rs");
     t.compile_fail("tests/ui/duplicate_pymodule_submodule.rs");
+    t.compile_fail("tests/ui/invalid_base_class.rs");
```
Comment (davidhewitt) [inline tests/test_compile_error.rs:65]: Given the diagnostic contains extra detail on `abi3`:

```suggestion
    #[cfg(not(Py_LIMITED_API))]
    t.compile_fail("tests/ui/invalid_base_class.rs");
```

**Comment by ChayimFriedman2** [comment]
@davidhewitt How come this PR decreased coverage when all it did was fiddling with types? Do you have an idea?

**Comment by davidhewitt** [comment]
Pretty sure the coverage is due to other changes reducing coverage on `main` compared to the merge base on this PR. I think it can be fixed by #4472.

Let's merge this - thanks very much!

**Comment by ChayimFriedman2** [comment]
@davidhewitt GitHub doesn't agree to merge this because of said coverage regression...

File: tests/ui/invalid_base_class.stderr [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  |
+  = help: the trait `PyClassBaseType` is not implemented for `PyBool`
+  = note: if you own `PyBool`, add `subclass` to the `#[pyclass]` macro: `#[pyclass(subclass)]`
+  = help: the following other types implement trait `PyClassBaseType`:
+            PyAny
+            PyArithmeticError
+            PyAssertionError
+            PyAttributeError
+            PyBaseException
+            PyBaseExceptionGroup
```
Comment (davidhewitt) [inline tests/ui/invalid_base_class.stderr:15]: The problem is that `BaseExceptionGroup` isn't defined until 3.11 (IIRC?).

You'll need to put a minimum Python version on this test, I think.

Comment (ChayimFriedman2) [reply tests/ui/invalid_base_class.stderr:15]: Fixed that, thanks.

**Review by davidhewitt — APPROVED** [review]

**Comment by davidhewitt** [comment]
I think the `test-debug` failure is another manifestation of https://github.com/PyO3/pyo3/pull/4185#issuecomment-2113722396 🤔

File: src/impl_/pyclass.rs [hunk 3]
```diff
@@ -1116,7 +1116,18 @@ impl<T> PyClassThreadChecker<T> for ThreadCheckerImpl {
...
+        note = "if you own `{Self}`, add `subclass` to the `#[pyclass]` macro: `#[pyclass(subclass)]`",
+        note = "with the `abi3` feature enabled, PyO3 does not support subclassing native types",
+    )
+)]
+#[cfg_attr(
+    all(diagnostic_namespace, not(feature = "abi3")),
+    diagnostic::on_unimplemented(
+        message = "pyclass `{Self}` cannot be subclassed",
+        label = "required for `#[pyclass(extends={Self})]`",
+        note = "if you own `{Self}`, add `subclass` to the `#[pyclass]` macro: `#[pyclass(subclass)]`",
```
Comment (mejrs) [inline src/impl_/pyclass.rs:1130]: Compiler error suggestions tend to have the form of "to do x, use y", not "do x". I think we ought to do the same.


```suggestion
        note = "`{Self}` must have `#[pyclass(subclass)]` to be eligible for subclassing"
```
(same for the other one)

**Comment by ChayimFriedman2** [comment]
I can't reproduce the CI failure locally. Why does it even run the test under abi3 when it is explicitly cfg'ed out in this case?

**Comment by davidhewitt** [comment]
It's a mismatch between how the configuration and features are set up for that job. I have an idea how to make progress here that I'm going to play with tonight.

**Comment by davidhewitt** [comment]
@ChayimFriedman2 I pushed a commit which should fix the `test-debug` issue. Looks like I got clippy wrong; don't worry about fixing here. If `test-debug` passes we can drop the commit from this branch and merge it in #4497.

**Comment by davidhewitt** [comment]
I merged #4497 so I dropped the commit from here; hopefully CI is green and we can merge this shortly!

**Review by davidhewitt — APPROVED** [review]
Thanks again!

**Pull request merged**